File: hostel-management-backend/src/main/java/com/iiitb/hostel/service/CheckService.java
package com.iiitb.hostel.service;

import com.iiitb.hostel.dto.CheckInOutRequest;
import com.iiitb.hostel.dto.CheckPreviewDto;
import com.iiitb.hostel.model.Allocation;

public interface CheckService {

    CheckPreviewDto getCheckInPreview(String rollNumber);

    Allocation performCheckIn(CheckInOutRequest request, String performedBy);

    CheckPreviewDto getCheckOutPreview(String rollNumber);

    Allocation performCheckOut(CheckInOutRequest request, String performedBy);
}

File: hostel-management-frontend/src/services/checkService.ts
import api from "./api";
import type { Allocation, Student } from "../types";

export interface CheckPreviewDto {
  student: Student;
  allocation: Allocation;
}

export const getCheckInPreview = async (rollNumber: string): Promise<CheckPreviewDto> => {
  const encoded = encodeURIComponent(rollNumber.trim());
  const response = await api.get<CheckPreviewDto>(`/check/in/${encoded}`);
  return response.data;
};

export const performCheckIn = async (payload: { rollNumber: string; remarks?: string }): Promise<unknown> => {
  const response = await api.post("/check/in", payload);
  return response.data;
};

export const getCheckOutPreview = async (rollNumber: string): Promise<CheckPreviewDto> => {
  const encoded = encodeURIComponent(rollNumber.trim());
  const response = await api.get<CheckPreviewDto>(`/check/out/${encoded}`);
  return response.data;
};

export const performCheckOut = async (payload: { rollNumber: string; remarks?: string }): Promise<unknown> => {
  const response = await api.post("/check/out", payload);
  return response.data;
};
